mods/cnc/maps/gdi09/gdi09-AI.lua

```diff
@@ -0,0 +1,201 @@
+	end)
+end
+
+CheckForHarvester = function()
+	local harv = Nod.GetActorsByType("harv")
+	return #harv > 0
+end
+
+GuardBase = function()
+	Utils.Do(NodBase, function(building)
+		Trigger.OnDamaged(building, function()
+			Utils.Do(NodGuards, function(guard)
+				if not guard.IsDead and not building.IsDead then
+					guard.Stop()
+					guard.Guard(building)
+				end
+			end)
+		end)
+	end)
+end
+
+ProduceHarvester = function(building)
+	if not buildingHarvester then
+		buildingHarvester = true
+		building.Build(HarvesterProductionType, function()
+			buildingHarvester = false
+		end)
+	end
+end
+
+ProduceInfantry = function(building)
+	if building.IsDead or building.Owner ~= Nod then
+		return
+	elseif not CheckForHarvester() then
+		Trigger.AfterDelay(DateTime.Seconds(10), function() ProduceInfantry(building) end)
+		return
+	end
+
+	if #PatrolProductionQueue >= 1 then
+		local inQueue = PatrolProductionQueue[1]
+		local toBuild =  { inQueue.unit[1] }  
+		local patrolPath = inQueue.waypoints
+		building.Build(toBuild, function(unit)
+			ReplenishPatrolUnit(unit[1], handofnod, patrolPath, 40)
+			table.remove(PatrolProductionQueue, 1)
+		end)
+		Trigger.AfterDelay(DateTime.Seconds(10), function() ProduceInfantry(building) end)
+		return
+	end
+
+	local delay = Utils.RandomInteger(DateTime.Seconds(3), DateTime.Seconds(9))
+	local toBuild = { Utils.Random(InfantryProductionTypes) }
+	local Path = Utils.Random(AttackPaths)
+	building.Build(toBuild, function(unit)
+		InfantryAttackGroup[#InfantryAttackGroup + 1] = unit[1]
+
+		if #InfantryAttackGroup >= InfantryGroupSize then
+			MoveAndHunt(InfantryAttackGroup, Path)
+			InfantryAttackGroup = { }
+			Trigger.AfterDelay(InfantryProductionCooldown, function() ProduceInfantry(building) end)
+		else
+			Trigger.AfterDelay(delay, function() ProduceInfantry(building) end)
+		end
+	end)
+
+end
+
+ProduceVehicle = function(building)
+	if building.IsDead or building.Owner ~= Nod then
+		return
+	elseif not CheckForHarvester() then
+		ProduceHarvester(building)
+		Trigger.AfterDelay(DateTime.Seconds(10), function() ProduceVehicle(building) end)
+		return
+	end
+
+	local delay = Utils.RandomInteger(DateTime.Seconds(12), DateTime.Seconds(17))
+	local toBuild = { Utils.Random(VehicleProductionTypes) }
+	local Path = Utils.Random(AttackPaths)
+	building.Build(toBuild, function(unit)
+		VehicleAttackGroup[#VehicleAttackGroup + 1] = unit[1]
+
+		if #VehicleAttackGroup >= VehicleGroupSize then
+			MoveAndHunt(VehicleAttackGroup, Path)
+			VehicleAttackGroup = { }
+			Trigger.AfterDelay(VehicleProductionCooldown, function() ProduceVehicle(building) end)
+		else
+			Trigger.AfterDelay(delay, function() ProduceVehicle(building) end)
+		end
+	end)
+end
+
+StartAI = function()
+	RepairNamedActors(Nod, 0.75)
+
+	Nod.Cash = StartingCash
+	GuardBase()
+end
+
+Trigger.OnAllKilledOrCaptured(NodBase, function()
+	Utils.Do(GDI.GetGroundAttackers(), IdleHunt)
+end)
+
+rigger.OnKilled(nodrefinery, function(building)
```

Comment:
This probably want to be `Trigger`?

Reply:
Thanks.
I was also thinking about replacing those triggers with:

```
...
BaseNuke5 = { type = "nuke", pos = CPos.New(12, 30), cost = 500 }
InfantryProduction = { type = "hand", pos = CPos.New(15, 24), cost = 500 }
VehicleProduction = { type = "afld", pos = CPos.New(3, 27), cost = 2000 }
RebuildBaseOrder = { 
	{ actor = nodrefinery, replacement = BaseRefinery },
	{ actor = nodpower1, replacement = BaseNuke1 },
	{ actor = nodpower2, replacement = BaseNuke2 },
	{ actor = nodpower3, replacement = BaseNuke3 },
	{ actor = nodpower4, replacement = BaseNuke4 },
	{ actor = nodpower5, replacement = BaseNuke5 },
	{ actor = handofnod, replacement = InfantryProduction },
	{ actor = nodairfield, replacement = VehicleProduction },
}
...
RebuildBase = function()
	for k, order in pairs(RebuildBaseOrder) do
		Trigger.OnKilled(order.actor, function()
			BuildBuilding(order.replacement, NodCYard)
		end)
	end
	RebuildBaseOrder = nil
end
```
It may be a bit more clean but I still don't love it.